ios/storage.md

```diff
@@ -732,25 +611,60 @@ After adding the new mutation in our operations file, we run the code generator
 
-By default, all Amazon S3 resources are private. If you want your users to have access to Amazon S3 buckets
-or objects, you can assign appropriate permissions with an [IAM policy](http://docs.aws.amazon.com/IAM/latest/UserGuide/PoliciesOverview.html).
+```swift
+extension S3ObjectInput: AWSS3ObjectProtocol, AWSS3InputObjectProtocol {
+    // ...
+}
 
-However, what if you wanted to provide permissions temporarily, for example: _you want to share a link to file temporarily and have the link expire after a set time_. To do this using an IAM policy would require you to first setup the policy to grant access and then at a later time remember to delete the IAM policy to revoke access. 
+extension AWSS3PreSignedURLBuilder: AWSS3ObjectPresignedURLGenerator  {
+    // ...
+}
+
+extension AWSS3TransferUtility: AWSS3ObjectManager {
+    // ...
+}
+```
+
+There is an implementation in the iOS Test Suite that can be used as a reference: [aws-mobile-appsync-sdk-ios/AWSAppSyncTestCommon/S3ObjectWrapper.swift](https://github.com/awslabs/aws-mobile-appsync-sdk-ios/blob/master/AWSAppSyncTestCommon/S3ObjectWrapper.swift). That code can be used as drop-in implementation for most use cases.
+
+The mutation operation doesn't require any specific changes in method signature. It requires only an S3ObjectInput with `bucket`, `key`, `region`, `localUri`, and `mimeType`. Now when you do a mutation, it automatically uploads the specified file to Amazon S3 using the `AWSS3TransferUtility` client internally.
 
-Alternatively, you can use pre-signed URLs to give your users temporary access to Amazon S3 objects.  When you create a pre-signed URL, you must provide your security credentials, specify a bucket name, an object key, an HTTP method, and an expiration date and time. The pre-signed URL is valid only for the specified duration.
+## Working with access levels in your app code
+
+By default, all Amazon S3 resources are private. If you want your users to have access to Amazon S3 buckets or objects, you can assign appropriate permissions with an [IAM policy](http://docs.aws.amazon.com/IAM/latest/UserGuide/PoliciesOverview.html).
+
+### IAM Policy Based Permissions
+
+When you upload objects to the S3 bucket the Amplify CLI creates, you must manually prepend the appropriate access-level information to the `key`. The correct prefix - `public/`, `protected/` or `private/` - will depend on the access level of the object as documented in the [Storage Access section](./storage#storage-access). For example:
+
+```swift
+var s3Object = S3ObjectInput()
+// Accessible by all users
+s3Object.key = "public/myFile.txt"
+// Readable by all users, but writable only by the creating user
+s3Object.key = "protected/\(cognitoIdentityId)/myFile.txt"
```

Comment:
Could we put a link on how to get the IdentityID from AWSMobileClient? https://aws-amplify.github.io/docs/ios/authentication#utility-properties

Reply:
good call, done